diff -x puta -x '*.img' -ru 023_spin_lock/Makefile 024_exec/Makefile
--- 023_spin_lock/Makefile	2019-06-23 21:00:40.062171779 +0900
+++ 024_exec/Makefile	2019-06-23 21:00:40.094171516 +0900
@@ -2,7 +2,7 @@
 CFLAGS = -Wall -Wextra -nostdinc -nostdlib -fno-builtin -fno-common -Iinclude
 LDFLAGS = -Map kernel.map -s -x -T kernel.ld
 OBJS = main.o syscall.o proc.o iv.o fbcon.o fb.o font.o kbc.o x86.o	\
-	intr.o pic.o apic.o sched.o hpet.o acpi.o nic.o pci.o		\
+	ap.o intr.o pic.o apic.o sched.o hpet.o acpi.o nic.o pci.o	\
 	handler.o fs.o common.o
 
 $(TARGET): $(OBJS)
024_exec のみに存在: ap.c
024_exec/include のみに存在: ap.h
diff -x puta -x '*.img' -ru 023_spin_lock/main.c 024_exec/main.c
--- 023_spin_lock/main.c	2019-06-23 21:00:40.074171681 +0900
+++ 024_exec/main.c	2019-06-23 21:00:40.126171254 +0900
@@ -1,4 +1,5 @@
 #include <x86.h>
+#include <ap.h>
 #include <intr.h>
 #include <pic.h>
 #include <apic.h>
@@ -31,17 +32,10 @@
 {
 	unsigned char pnum = get_pnum();
 
-	/* APの場合、初期化処理をスキップ */
+	/* 専用の初期化処理を行い、実行を開始する */
 	if (pnum) {
-		/* コンソールの初期化が完了するまで待つ */
-		while (!is_con_inited);
-
-		/* AP自身のプロセッサ番号を表示 */
-		puth(pnum, 1);
-
-		/* haltして待つ */
-		while (1)
-			cpu_halt();
+		ap_init();
+		ap_run(pnum);
 	}
 
 	/* フレームバッファ周りの初期化 */
@@ -51,13 +45,6 @@
 	clear_screen();
 	is_con_inited = 1;
 
-	/* BSP自身のプロセッサ番号を表示 */
-	puth(pnum, 1);
-
-	/* haltして待つ */
-	while (1)
-		cpu_halt();
-
 	/* ACPIの初期化 */
 	acpi_init(pi->rsdp);
 
@@ -77,6 +64,16 @@
 	/* ファイルシステムの初期化 */
 	fs_init(_fs_start);
 
+	/* AP1で外部アプリ実行 */
+	struct file *app = open("puta");
+	ap_enq_task(app, 1);	/* 1回目 */
+	while (ap_enq_task(app, 1) != 0) /* 2回目を試みる */
+		CPU_PAUSE();
+
+	/* haltして待つ */
+	while (1)
+		cpu_halt();
+
 	/* スケジューラの初期化 */
 	sched_init();
 
